Use `uv` for development and packaging

linozen opened the pull request:
This is just so I can work effectively on the project. I added `onnxruntime-gpu` as an extra, called `cuda_gpu`.

I also had to do some minor refactoring (rename `__main__.py` to `cli.py`) to make everything work with `uv`. Tomorrow, I will get to work on Mac OS X integration. Currently, I'm testing [pywhispercpp](https://github.com/absadiki/pywhispercpp)

gaspardpetit commented:
Thanks! I'll take a look - for now I have concerns with some of the changes to the project structure. I see you added a `pyproject.toml` - this is fine, but I don't think we should keep the `setup.py` then.

The `uv.lock` is new to me as well - is this a poetry lock file? I don't mind pinning down dependency versions in a lock file. At the moment, I am just using `venv` - so I'd probably  just go for `pip freeze > requirements.lock`

For cli.py vs main.py - I'm ok with the change, but I suggest we keep a __main__.py for providing an entry point when calling `py -m verbatim`

Feel free to challenge anything I propose here !

linozen commented:
> Thanks! I'll take a look - for now I have concerns with some of the changes to the project structure. I see you added a `pyproject.toml` - this is fine, but I don't think we should keep the `setup.py` then.

Yes, that's right, the setup.py can be removed once the more modern pyproject.toml satisfies all requirements. 

> The `uv.lock` is new to me as well - is this a poetry lock file? I don't mind pinning down dependency versions in a lock file. At the moment, I am just using `venv` - so I'd probably just go for `pip freeze > requirements.lock`

Well, it's similar to poetry but in my estimation around 15x faster. The nice thing about `uv` is that it handles everything. Python binary version management, dependency locking, building and publishing a package - all in one tool that works very well and is a joy to use. After having tried many different Python packaging tools, I believe if this irons out some of the kinks (e.g. improve documentation in some places), I don't ever want to use something else again.

> For cli.py vs main.py - I'm ok with the change, but I suggest we keep a **main**.py for providing an entry point when calling `py -m verbatim`

I have no passion about this either way. Something that I could not quite pin done broke when I tried to invoke the project.script with `__main__.py`.

I'd suggest I'll polish this a bit more and then you can give it all a spin. Until then I really suggest you try out `uv`. https://docs.astral.sh/uv/

gaspardpetit commented:
I've been looking at ruff and so far I like it. I'm also liking the pyproject.toml structure you suggested and using `hatchling` is great. I'll give `uv` a try too. 

For versioning, I think you can do:

```
[project]
name = "verbatim"
dynamic = ["version"]

[tool.hatch.version]
path = "verbatim/__init__.py"
```

linozen commented:
Hey @gaspardpetit,

now, this is pretty much complete from my side:
- I renamed `cli.py` to `main.py` to clarify the entrypoint
- migrated all information to pyproject.toml and removed `setup.py`
- created one `requirements.txt` and one `requirements-all.txt` file for compatibility (e.g. docker etc.), the latter of which includes `onnxruntime-gpu`. Let me know if you want to have them automatically generated with a pre-commit hook. See https://docs.astral.sh/uv/guides/integration/pre-commit/
- Did one formatting run with `ruff formar verbatim/` from root dir.

linozen commented:
From my POV, this would be ready to merge. Let me know what you think.

gaspardpetit commented:
A few details to iron out - ex. automated checks are failing - but given the breadth of the changes I suggest we get them in and iron out from there. Thanks again !
